pkg/controllers/managementuser/rbac/roletemplates/crtb_handler.go

```diff
@@ -0,0 +1,145 @@
+package roletemplates
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+
+	v3 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
+	"github.com/rancher/rancher/pkg/rbac"
+	"github.com/rancher/rancher/pkg/types/config"
+	rbacv1 "github.com/rancher/wrangler/v3/pkg/generated/controllers/rbac/v1"
+	"github.com/rancher/wrangler/v3/pkg/name"
+	v1 "k8s.io/api/rbac/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	crtbOwnerLabel = "authz.cluster.cattle.io/crtb-owner"
+)
+
+type crtbHandler struct {
+	impersonationHandler *impersonationHandler
+	crbClient            rbacv1.ClusterRoleBindingController
+}
+
+func newCRTBHandler(uc *config.UserContext) *crtbHandler {
+	return &crtbHandler{
+		impersonationHandler: &impersonationHandler{
+			userContext: uc,
+			crClient:    uc.RBACw.ClusterRole(),
+			crtbClient:  uc.Management.Wrangler.Mgmt.ClusterRoleTemplateBinding(),
+			prtbClient:  uc.Management.Wrangler.Mgmt.ProjectRoleTemplateBinding(),
+		},
+		crbClient: uc.RBACw.ClusterRoleBinding(),
+	}
+}
+
+// OnChange ensures that the correct ClusterRoleBinding exists for the ClusterRoleTemplateBinding
+func (c *crtbHandler) OnChange(key string, crtb *v3.ClusterRoleTemplateBinding) (*v3.ClusterRoleTemplateBinding, error) {
+	if crtb == nil || crtb.DeletionTimestamp != nil {
+		return nil, nil
+	}
+
+	crb, err := buildClusterRoleBinding(crtb)
+	if err != nil {
+		return nil, err
+	}
+
+	ownerLabel := createCRTBOwnerLabel(crtb.Name)
+	currentCRBs, err := c.crbClient.List(metav1.ListOptions{LabelSelector: ownerLabel})
+	if err != nil {
+		return nil, err
+	}
+
+	// Find if there is a CRB that already exists and delete all excess CRBs
+	var matchingCRB *v1.ClusterRoleBinding
+	for _, currentCRB := range currentCRBs.Items {
+		if areClusterRoleBindingsSame(crb, &currentCRB) && matchingCRB == nil {
+			matchingCRB = &currentCRB
+			continue
+		}
+		if err := c.crbClient.Delete(currentCRB.Name, &metav1.DeleteOptions{}); err != nil {
+			return nil, err
+		}
+	}
+
+	// If we didn't find an existing CRB, create it
+	if matchingCRB == nil {
+		if _, err := c.crbClient.Create(crb); err != nil {
+			return nil, err
+		}
+	}
```

Comment:
Could this cause a race condition while access is temporarily lost due to removing CRB but the new ones were not created yet? Or what if the Creation fails for some reason?
For both reasons, I would move the cleanup to the end of the function (also, under which circumstances there could there be multiple CRBs that need to be removed?